[IMP] extract: update documentation (#3835)
odoo/documentation: content/developer/reference/extract_api.rst

===========
Extract API
===========

Odoo provides a service to automate the processing of documents of type **invoices**, **expenses**
or **resumes**.

The service scans documents using an :abbr:`OCR (Optical Character Recognition)` engine and then
uses :abbr:`AI(Artificial Intelligence)`-based algorithms to extract fields of interest such as the
total, due date, or invoice lines for *invoices*, the total, date, payment reference for *expenses*,
or the name, email, phone number for *resumes*.

> This service is a paid service. Each document processing will cost you one credit. Three different
> sized packs can be bought on

Reviewer: ```suggestion
This service is a paid service. Each document processing will cost you one credit.
Credits can be bought on
```
In the case where we change the packs (remove one or add one), there won't be an inconsistency between this page and the IAP services page.

`iap.odoo.com <https://iap.odoo.com/iap/in-app-services/259?sortby=date>`_.

You can either use this service directly in the Accounting, Expense, or Recruitment App or through
the API. The Extract API, which is detailed in the next section, allows you to integrate our
service directly into your own projects.

Overview
========

The extract API uses the JSON-RPC2_ protocol, its endpoint routes is located at
https://iap-extract.odoo.com.

Version
-------

The version of the Extract API is specified in the route.

> The latests versions are::

Reviewer: ```suggestion
The latests versions are:
```

uses :abbr:`AI(Artificial Intelligence)`-based algorithms to extract fields of interest such as the
total, due date, or invoice lines for *invoices*, the total, date, payment reference for *expenses*,
or the name, email, phone number for *resumes*.

This service is a paid service. Each document processing will cost you one credit.
Credits can be bought on
`iap.odoo.com <https://iap.odoo.com/iap/in-app-services/259?sortby=date>`_.

You can either use this service directly in the Accounting, Expense, or Recruitment App or through
the API. The Extract API, which is detailed in the next section, allows you to integrate our
service directly into your own projects.

Overview
========

The extract API uses the JSON-RPC2_ protocol, its endpoint routes is located at
https://iap-extract.odoo.com.

Version
-------

The version of the Extract API is specified in the route.

The latests versions are:
    - invoices: 122
    - expenses: 132
    - applicant: 102

Flow
----

The flow is the same for the three services. In the following examples, `__service__` can be
replaced by `invoice`, `expense` or `applicant` to use the appropriate service.

> 1. | Call `/api/extract/__service__/1/parse`_ to submit your invoices (one call for each invoice).

Author: I did replace all `/api/extract/__service__/1/` in favor of `extract_api/` as well, because it looks simpler.

        response. Some results can be unavailable in older versions. If not specified, the latest
        version will be used.
    ``documents`` (required)
        The document must be provided as a string in the ASCII encoding. The list should contain
        only one string. If multiple strings are provided only the first string corresponding to a
        pdf will be processed. If no pdf is found, the first string will be processed. This field
        is a list only for legacy reasons. The supported extensions are *pdf*, *png*, *jpg* and
        *bmp*.
    ``dbuuid`` (optional)
        Unique identifier of the Odoo database.
    ``webhook_url`` (optional)
        A webhook URL can be provided. An empty POST request will be sent to
        ``webhook_url/document_uuid`` when the result is ready.
    ``user_infos`` (optional)
        Information concerning the person sending the document to the extract service. It can be
        the client or the supplier (depending on the ``perspective``). This information is not
        required in order for the service to work but it greatly improves the quality of the result.

>         ``user_company_vat`` (optional)

Reviewer 2: ```suggestion
   .. rst-class:: o-definition-list

        ``user_company_vat`` (optional)
```

            "status_code": int,  // deprecated
            "status_msg": string,
            "document_uuid": string,
            // "document_id": int,  // deprecated
        }
    }

.. warning::
    The ``document_id`` field is deprecated and will be removed in the future. Please use
    ``document_uuid`` instead.

    Version introducing ``document_uuid``:
     - invoices: 122
     - expenses: 132
     - applicant: 102

.. note::
    The API does not actually use the JSON-RPC error scheme. Instead the API has its own error
    scheme bundled inside a successful JSON-RPC result.

Get results
===========

.. _/api/extract/__service__/1/get_result:

Routes
------

``/api/extract/__service__/1/get_result``

    - /api/extract/invoice/1/get_result
    - /api/extract/invoice/1/get_result_batch
    - /api/extract/expense/1/get_result
    - /api/extract/expense/1/get_result_batch
    - /api/extract/applicant/1/get_result
    - /api/extract/applicant/1/get_result_batch
    - |ss| /api/extract/invoice/1/get_results |se| (deprecated)
    - |ss| /iap/invoice_extract/get_result |se| (deprecated)
    - |ss| /iap/expense_extract/get_result |se| (deprecated)

Request
-------

> ``jsonrpc`` (required)

Reviewer 2: ```suggestion
.. rst-class:: o-definition-list

``jsonrpc`` (required)
```

    see JSON-RPC2_
``method`` (required)
    see JSON-RPC2_
``id`` (required)
    see JSON-RPC2_
``params``
    Dictionary containing the following content.

    ``version`` (required)
        |SAME_AS_PARSE|
    ``documents_uuids`` (required, replaces ``documents_ids``)
        The list of ``document_id`` for which you want to get the current parsing status.
    ``documents_ids`` (deprecated)
        The list of ``document_id`` for which you want to get the current parsing status.

.. code-block:: js

    {
        "jsonrpc": "2.0",
        "method": "call",
        "params": {
            "version": int,
            // "documents_ids": [],  // deprecated
            "documents_uuids": [],
        },
        "id": string,
    }

.. note::

>     The example shows the request to the ``/api/extract/invoice/1/get_result_batch`` route.

Reviewer 2: What example? :monocle_face:

Author: changed to code snippet

    You can use the endpoint ``/api/extract/invoice/1/get_result`` to get the result of a single
    document. In that case, you don't need to provide a list of ``document_uuids`` but a single
    ``document_uuid``.

Response
--------

When getting the results from the parse, the detected field vary a lot depending on the type of
document. Each response is a list of dictionaries, one for each document. The keys of the dictionary
are the name of the field and the value is the value of the field.

> ``jsonrpc``

Reviewer 2: ```suggestion
.. rst-class:: o-definition-list

``jsonrpc``
```

    see JSON-RPC2_
``id``
    see JSON-RPC2_
``result``
    Dictionary where each key is a document_id. For each ``document_id``

>     ``status``

Reviewer 2: ```suggestion
.. rst-class:: o-definition-list

    ``status``
```

        |SAME_AS_PARSE|
    ``status_code``
        |SAME_AS_PARSE|
    ``status_msg``
        |SAME_AS_PARSE|
    ``results``

>         Only present if the request is successful.

Reviewer 2: ```suggestion
        Only present if the request is successful.
        
        .. rst-class:: o-definition-list
```

                "status_msg": string,
                "results": [
                    {
                        "full_text_annotation": string,
                        "feature_1_name": feature_1_result,
                        "feature_2_name": feature_2_result,
                        ...
                    },
                    ...
                ]
            },
            "document_id_2": {
                "status": string,
                "status_code": int,  // deprecated
                "status_msg": string,
                "results": [
                    {
                        "full_text_annotation": string,
                        "feature_1_name": feature_1_result,
                        "feature_2_name": feature_2_result,
                        ...
                    },
                    ...
                ]
            },
            ...
        }
    }

.. note::
    The example shows the response from the ``/api/extract/invoice/1/get_result_batch`` route.
    When using the ``/api/extract/invoice/1/get_result`` route (note the singularity), the response
    will be the value of the key associated to the invoice.

.. warning::

>     result keys are strings despite the fact that the ``document_ids`` given in the request body are

Reviewer 2: ```suggestion
    Result keys are strings despite the fact that the ``document_ids`` given in the request body are
```

     invoice). On success, you receive a `document_uuid` in the response.
#. | You then have to regularly poll :ref:`/get_result <extract_api/get_result>` to get the
     document's parsing status.
   | Alternatively, you can provide a `webhook_url` at the time of the call to
     :ref:`/parse <extract_api/parse>` and you will be notified (via a POST request) when the
     result is ready.
#. | Once the result is received, you can validate it by calling
     :ref:`/validate <extract_api/validate>` and sending the expected values.
   | This step is optional but greatly helps the system to improve.

The HTTP POST method should be used for all of them. A python implementation of the full flow for
invoices can be found :download:`here <extract_api/implementation.py>` and a token for integration
testing is provided in the
:ref:`integration testing section <latestextract_api/integration_testing>`.

Parse
=====

Request the processing of a document from the OCR. The route will return a `document_uuid`
(it replaces `document_id`, which is deprecated); you can use it to obtain the result of your
request.

> .. _extract_api/parse:

Reviewer 2: I'm unsure if you intend to have this anchor target the "Routes" section or the parent section "Parse"? In the latter case, you must move the anchor before the section title.

Author: It is meant for the routes, since this is what it is referring to when used in the text.

   | Alternatively, you can provide a `webhook_url` at the time of the call to `/api/extract/__service__/1/parse`_ and
     you will be notified (via a POST request) when the result is ready.
3. | Once the result received, you can validate it by calling `/api/extract/__service__/1/validate`_ and sending the
     expected values.
   | This step is optional but greatly helps the system to improve.

The HTTP POST method should be used for all of them. A python implementation of the full flow for invoices can be found
:download:`here <extract_api/implementation.py>` and a token for integration testing is provided in the
:ref:`integration testing section <webservices/extract_api/integration_testing>`.

Parse
=====

Request the processing of a document from the OCR. The route will return a `document_uuid` (replaces `document_id` which
is deprecated), you can use to obtain the result of your request.

.. _/api/extract/__service__/1/parse:

Routes
------

``/api/extract/__service__/1/parse``

    - /iap/invoice_extract/parse (deprecated)
    - /api/extract/invoice/1/parse
    - /iap/expense_extract/parse (deprecated)
    - /api/extract/expense/1/parse

>     - /api/extract/applicant/1/parse

Author: perfect!

Request
-------

``jsonrpc`` (required)
    Must be “2.0”.
``method`` (required)
    Must be “call”.
``id`` (required)
    An identifier established by the client. It allows the client to keep track of which response goes with which

>     request. This makes asynchronous calls easier.

Author: Replaced the content by `see JSON-RPC2_`.

``params``
    ``account_token`` (required)
        The token of the account from which credits will be taken. Each successful call costs one token.
    ``version`` (optional)
        The version will determine the format of your requests and the format of the server response. Some results can be

>         unavailable in older versions. If not specified, the latest version will be used.

Author: I added it to the `Overview`

    ``status_msg``
        |SAME_AS_PARSE|

========================  ===========================  ==========================================================
status                    status_code (deprecated)     status_msg
========================  ===========================  ==========================================================
success                   0                            Success
error_validation_format   12                           Validation format is incorrect
========================  ===========================  ==========================================================

.. code-block:: js

    {
        "jsonrpc": "2.0",
        "id": string,
        "result": {
            "status": string,
            // "status_code": int,  // deprecated
            "status_msg": string,
        }
    }

.. _webservices/extract_api/integration_testing:

Integration Testing
===================

You can test your integration by using *integration_token* as ``account_token`` in the
:ref:`/parse </api/extract/__service__/1/parse>` request.

Using this token put you in test mode and allows you to simulate the entire flow without really parsing a document and

> without being billed one credit for each successful **invoice** parsing.

Author: This one was on purpose, but if we make integration testing available to the other types of documents then it makes sense.

The only technical differences in test mode is that the document you send is not parsed by the system and that the

> response you get from :ref:`/get_results </api/extract/__service__/1/get_results>` is a hard-coded one.

Author: This was just because of the ref... I can just edit what is visible since I used the same anchor for both routes.
```suggestion
response you get from :ref:`/get_result </api/extract/__service__/1/get_results>` is a hard-coded one.
```

> response you get from :ref:`/get_results </api/extract/__service__/1/get_results>` is a hard-coded one.

Reviewer: Ok, I see, then we might want to also change the anchor to `get_result`, no ?
